sql: fix dangling zone config on REGIONAL BY ROW -> REGIONAL BY TABLE

otan opened the pull request:
We previously did not set NumReplicas to 0 for RBR -> RBT conversions.
This means the zone configuration do not cleanup and inherit properly
after the async job completes for dropping old indexes since it no
longer thinks the RBR zone config is a placeholder, leaving a
dangling reference. Patch this up correctly.

N.B.: In theory there's nothing wrong with this. But this patch makes
SHOW ZONE CONFIGURATION consistent when you convert RBR to RBT
compared to making a RBT table directly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60659)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
Nit: Not following what you mean by "...since it no longer things the RBR zone config is a placeholder".  Can you clarify?  

Also, last sentence may be clearer if you say "But this patch ensures that SHOW ZONE CONFIGURATION will return the same result for tables created RBT as those that have been altered from RBR to RBT."

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60659#-:-MTpAt5t6SS1GSPXWKQu:b-s9vz0p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/region_util.go, line 563 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60659#-MTpDThECmILKQklP12f:-MTpDThECmILKQklP12g:b-si1whm) ([raw file](https://github.com/cockroachdb/cockroach/blob/05ce19cf0276541592d4e007dbef5a9cc7168dd2/pkg/sql/region_util.go#L563)):*
> ```Go
> 	// Mark the NumReplicas as 0 if we have subzones but no other features
> 	// in the zone config. This signifies a placeholder.
> 	if len(zoneConfig.Subzones) > 0 && isPlaceholderZoneConfigForMultiRegion(zoneConfig) {
> ```

Nit: in this check, can we not use "hasNewSubzones"?  If there are existing subzones, will the NumReplicas not already be set to 0?


<!-- Sent from Reviewable.io -->

otan committed:
sql: fix dangling zone config on REGIONAL BY ROW -> REGIONAL BY TABLE

We previously did not set NumReplicas to 0 for RBR -> RBT conversions.
This means the zone configuration do not cleanup and inherit properly
after the async job completes for dropping old indexes since it no
longer thinks the RBR zone config is a placeholder, leaving a
dangling reference. Patch this up correctly.

N.B.: In theory there's nothing wrong with this. But this patch makes
SHOW ZONE CONFIGURATION consistent when you convert RBR to RBT
compared to making a RBT table directly.

Release note: None

otan reviewed (commented):
Oops typo. see the new commit msg!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60659#-:-MTqbh52A0cRdEeOVDgn:bgxf5fj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/sql/region_util.go, line 563 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60659#-MTpDThECmILKQklP12f:-MTqbZIv7On70nCSktwa:b-kfn4ik) ([raw file](https://github.com/cockroachdb/cockroach/blob/05ce19cf0276541592d4e007dbef5a9cc7168dd2/pkg/sql/region_util.go#L563)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: in this check, can we not use "hasNewSubzones"?  If there are existing subzones, will the NumReplicas not already be set to 0?
</blockquote></details>

nah -- added a comment. basically, we can have no new subzones but subzones still one the zoneconfig in a RBR - > RBT migration


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm
